src/OmniSharp/Api/v1/Highlighting/OmnisharpController.Highlighting.cs

```diff
@@ -0,0 +1,97 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Mvc;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Text;
+using OmniSharp.Models;
+
+namespace OmniSharp
+{
+    public partial class OmnisharpController
+    {
+        [HttpPost("highlight")]
+        public async Task<IEnumerable<HighlightResponse>> Highlight(HighlightRequest request)
+        {
+            var documents = _workspace.GetDocuments(request.FileName);
+            if (request.ProjectNames != null && request.ProjectNames.Length > 0)
+            {
+                documents = documents.Where(d => request.ProjectNames.Contains(d.Project.Name, StringComparer.Ordinal));
+            }
+
+            var results = new List<ClassifiedResult>();
+
+            foreach (var document in documents)
+            {
+                var project = document.Project.Name;
+                var text = await document.GetTextAsync();
+                var spans = new List<ClassifiedSpan>();
+
+                if (request.Lines == null || request.Lines.Length == 0)
+                {
+                    foreach (var span in await Classifier.GetClassifiedSpansAsync(document, new TextSpan(0, text.Length)))
+                    {
+                        spans.Add(span);
+                    }
+                }
+                else
+                {
+                    foreach (var line in request.Lines)
+                    {
+                        foreach (var span in await Classifier.GetClassifiedSpansAsync(document, text.Lines[line - 1].Span))
+                        {
+                            spans.Add(span);
+                        }
+                    }
+                }
+
+                results.AddRange(FilterSpans(request, spans)
+                    .Select(span => new ClassifiedResult()
+                    {
+                        Span = span,
+                        Lines = text.Lines,
+                        Project = project
+                    }));
+            }
+
+            return results
+                .GroupBy(z => z.Span.TextSpan.ToString())
+                .Select(a => HighlightResponse.FromClassifiedSpan(a.First().Span, a.First().Lines, a.Select(z => z.Project)));
+        }
+
+        class ClassifiedResult
+        {
+            public ClassifiedSpan Span { get; set; }
+            public TextLineCollection Lines { get; set; }
+            public string Project { get; set; }
+        }
+
+        private IEnumerable<ClassifiedSpan> FilterSpans(HighlightRequest request, IEnumerable<ClassifiedSpan> spans)
```

Comment:
This could could pretty much disappear if we put the classification types in the request.